### PR TITLE
feat: stubs cleanup

### PR DESCRIPTION
**Pull request by aarnphm**
Signed-off-by: Aaron Pham <29749331+aarnphm@users.noreply.github.com>

introduces python-minifier to reduce stubs size
from
image

to
image

### CODE REVIEW

File: scripts/tools/stubs_cleanup.sh [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+GIT_ROOT=$(git rev-parse --show-toplevel)
+MINIFY_OPTS=( --remove-literal-statements --no-remove-annotations --no-hoist-literals --no-rename-locals --no-remove-object-base --no-convert-posargs-to-args )
+
+cd "$GIT_ROOT" || exit 1
+
+source ./scripts/ci/helpers.sh
+
+need_cmd vim || FAIL "You will need to install vim to use this script, or use any editor of choice\nrequires to set \$EDITOR=<your_text_editor>"
+PROCESSED_TXT="$GIT_ROOT"/typings/processed.txt
+EDITOR=$(echo "$EDITOR") || /usr/bin/vim
```
[the remark this answers is not in the document]
Comment (aarnphm) [reply scripts/tools/stubs_cleanup.sh:12]: that make sense

File: scripts/tools/stubs_cleanup.sh [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+for file in $(git ls-files | grep -e "**.pyi$"); do
+  if [ ! -z $(grep "$file" "$PROCESSED_TXT") ]; then
+    PASS "$file already minified, skipping..."
+    continue
+  else
+    INFO "Processing $file..."
+    INFO "Removing pyright bugs..."
+    sed -i "s/],:/]/g" "$file"
+    sed -i "s/,,/,/g" "$file"
+    sed -i "s/]\n    .../]: .../g" "$file"
```
[the remark this answers is not in the document]
Comment (aarnphm) [reply scripts/tools/stubs_cleanup.sh:31]: ugh sometimes the parameter is missing from the stubs

File: scripts/tools/stubs_cleanup.sh [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    # sed -i "s/]$/]:/g" "$file"
+    cp "$file" "$file".bak && rm "$file"
+    if ! pyminify "${MINIFY_OPTS[@]}" "$file".bak &> "$file"; then
```
[the remark this answers is not in the document]
Comment (aarnphm) [reply scripts/tools/stubs_cleanup.sh:34]: yes

File: scripts/tools/stubs_cleanup.sh [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+      cp "$file" "$file".bak
+      if ! pyminify "${MINIFY_OPTS[@]}" "$file".bak &> "$file"; then
+        FAIL "Failed to fix $files. One can also use https://python-minifier.com/ to test where the problem may be. Make sure to match ${MINIFY_OPTS[@]}\nExitting now..."
+        rm "$file"
+        mv "$file".bak "$file"
+        exit 1
+      fi
+    fi
+    black --fast --config "$GIT_ROOT"/pyproject.toml --pyi "$file"
+    printf "%s\n" "$file" >> "$PROCESSED_TXT"
```
[the remark this answers is not in the document]
Comment (aarnphm) [reply scripts/tools/stubs_cleanup.sh:48]: ;(

File: scripts/tools/stubs_cleanup.sh [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    /usr/bin/rm "$file".bak
```
[the remark this answers is not in the document]
Comment (aarnphm) [reply scripts/tools/stubs_cleanup.sh:49]: it should be `\rm` instead, sometimes users can have different rm comments

File: scripts/tools/stubs_cleanup.sh [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+  else
+    INFO "Processing $file..."
+    INFO "Removing pyright bugs..."
+    sed -i "s/],:/]/g" "$file"
+    sed -i "s/,,/,/g" "$file"
+    sed -i "s/]\n    .../]: .../g" "$file"
+    # sed -i "s/]$/]:/g" "$file"
+    cp "$file" "$file".bak && rm "$file"
+    if ! pyminify "${MINIFY_OPTS[@]}" "$file".bak &> "$file"; then
+      FAIL "unable to processed $file, reverting to previous state, opening editor to fix..."
```
[the remark this answers is not in the document]
Comment (aarnphm) [reply scripts/tools/stubs_cleanup.sh:35]: sure we can do that as well, I think it is pretty neat for it to open the editor and let users fix it right away

File: scripts/tools/stubs_cleanup.sh [hunk 7]
```diff
@@ -0,0 +1,49 @@
+#!/usr/bin/env bash
+
+GIT_ROOT=$(git rev-parse --show-toplevel)
+MINIFY_OPTS=( --remove-literal-statements --no-remove-annotations --no-hoist-literals --no-rename-locals --no-remove-object-base --no-convert-posargs-to-args )
+
+cd "$GIT_ROOT" || exit 1
+
+source ./scripts/ci/helpers.sh
+
+need_cmd "$EDITOR" || FAIL "You will need to install vim to use this script, or use any editor of choice\nrequires to set \$EDITOR=<your_text_editor>"
```
Comment (parano) [inline scripts/tools/stubs_cleanup.sh:10]:
```suggestion
need_cmd "$EDITOR" || FAIL "You will need an editor to run this script, set your editor with env var: \$EDITOR=<your_text_editor>"
```

File: scripts/tools/stubs_cleanup.sh [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+PROCESSED_TXT="$GIT_ROOT"/typings/processed.txt
+: "${EDITOR:=/usr/bin/nano}"
+
+if [[ ! -f "$PROCESSED_TXT" ]]; then
+  touch "$PROCESSED_TXT"
+fi
+
+need_cmd black || (echo "make sure to run \`make install-dev-deps\`"; exit 1);
```
Comment (parano) [inline scripts/tools/stubs_cleanup.sh:18]:
```suggestion
need_cmd black || (echo "black command not found, install dev dependencies with \`make install-dev-deps\`"; exit 1);
```

File: scripts/tools/stubs_cleanup.sh [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+
+for file in $(git ls-files | grep -e "**.pyi$"); do
+  if [ ! -z $(grep "$file" "$PROCESSED_TXT") ]; then
+    PASS "$file already minified, skipping..."
+    continue
+  else
+    INFO "Processing $file ..."
+    INFO "Removing pyright bugs..."
+    sed "${SED_OPTS[@]}" "s/],:/]/g; s/,,/,/g; s/]\\\n    .../]: .../g" "$file"
+    # sed -i "s/]$/]:/g" "$file"
```
Comment (parano) [inline scripts/tools/stubs_cleanup.sh:36]:
```suggestion
```

**Pull request merged**